### ISSUE TITLE
pubsub: implement streaming pull receipt

### ISSUE THREAD

**Issue by pongad** [GoogleCloudPlatform/google-cloud-ruby#1800]
StreamingPull currently has a "protocol" level bug affecting all languages.

TLDR: client libs should send modify ack deadline for all messages upon receiving them.

## Problem description
When streaming, `StreamingPullResponse` can be buffered, either by gRPC itself or by proxy/firewall.
There might be significant delay between the server sending a response and the client library reading it.
For this reason, by the time client library decides it should send a modify ack deadline ("modack"), it might be too late. The message might have already expired and resent.

Experimenting with Java client, we have seen many messages delivered 3-5 times because of this problem.

## What server-side is doing
To fix this, pubsub team is implementing the "receipt" feature.
The server will not start the expiration timer until it receives a receipt from the client [1]. A receipt here is similar to an ack. A receipt notifies the server that the client has *seen* the message, while an ack notifies that the client has finished with it.
This should make the client and server start expiration timer closer to each other.

[1] There's actually a limit on how long the server is willing to wait, so that a message isn't lost forever if the client doesn't send a receipt.

## What the client needs to do
The client should be modified to send these receipts.
For compatibility, these "receipts" are just modack requests. The client should modack messages after seeing it, in addition to what it is currently doing. These modacks can be batched for performance.

## Few more details for the curious
These modacks/receipts aren't strictly necessary for avoiding duplication, but it will help pubsub server expire messages on time. Without them, messages might expire too late.

Each `StreamingPullResponse` can contain multiple messages. In an implementation about to roll out, a modack for any message in the response will serve as receipts for all messages in that response. Still, client libs are encouraged to actually modack all messages since this behavior might change in the future.

**Comment by blowmage** [comment]
The streaming client does modack received messages as part of the leasing mechanism now. Are you saying that the client should modack request as soon as a response is received? If so, does making a modack request as soon as a response is received somewhat negate the benefits of leasing received messages and batching all acks, nacks, and leased modacks requests?

For example, if a stream received a response ever ~2 seconds, and the stream has a stream_ack of ~60 seconds, we would expect to receive 30 responses and make 1 request every minute. But if we modack every response, we will receive 30 responses and make 31 requests every minute.

**Comment by pongad** [comment]
> Are you saying that the client should modack request as soon as a response is received?

Yes, otherwise messages might expire too late.

> If so, does making a modack request as soon as a response is received somewhat negate the benefits of leasing received messages and batching all acks, nacks, and leased modacks requests?

Hmm. When user code finishes with a message, does Ruby impl wait until the message is about to expire before actually sending the ack? When we worked on Java and Go (sorry, I'm most familiar with those two), pubsub team encouraged us to send acks and nacks promptly, so that users don't loose work if the machine crashes etc. So, Java and Go were sending ~30 requests per minute anyway. To send modacks, we plan to just "sneak" them into the acks and nacks we're already sending.

**Comment by blowmage** [comment]
User-driven acks, nacks, and modacks are sent at the same cadence and lease modacks, in the same batch request. So if the stream specifies a 60 second ack, then that batch request is sent about once a minute.

@lukesneeringer Thoughts on this?

**Comment by lukesneeringer** [comment]
Would it be possible to have an entirely separate modack thread that only handles the newly-received messages and sends on a much faster cadence (say, a second)?

**Comment by pongad** [comment]
Doing that is probably OK. If one second turns out to be too long, I suppose it's not much work to tweak the time later.

That said, my feeling is that if you're going to modack anyway, you might as well ack and nack too. @mdietz94 Do you have an opinion?

**Comment by mdietz94** [comment]
I think what Java/Go are doing is preferred, there's no reason to avoid ack/nack if you are sending a request anyway, and there's no reason to tie ack/nack to stream_ack_deadline. With this change, stream_ack_deadline is effectively meaningless anyway, since we will not expire until we receive a receipt, so it is best to just replace the timer there currently is with 1s, and do everything every second.

1s is not too long to batch receipts, that's a reasonable setting.

**Comment by blowmage** [comment]
It turns out we already had separate lease time and a much shorter ack/modack time. My apologies for being confused on this earlier.

I believe #1880 satisfies this request. PTAL.
